File: LeetCode/twopoint/LongestPalindromicSubstring.java
package twopoint;

public class LongestPalindromicSubstring {
    public String longestPalindrome(String s) {
        if (s.length() <= 1) {
            return s;
        }

        String result = "";

        for (int i = 0; i < s.length(); i++) {
            String oddString = palindromeFromCenter(s, i, i); // Odd length palindrome
            String evenString = palindromeFromCenter(s, i, i + 1); // Even length palindrome

            result = result.length() > oddString.length() ? result : oddString;
            result = result.length() > evenString.length() ? result : evenString;
        }

        return result;
    }

    private String palindromeFromCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        // left + 1 because when the chars are same, left--, so need to +1 to the right spot
        // right stay the same because substring is [ )
        return s.substring(left + 1, right);
    }

    public static void main(String[] args) {
        LongestPalindromicSubstring solution = new LongestPalindromicSubstring();

        // Test case
        String s = "babad";
        System.out.println("Longest palindromic substring: " + solution.longestPalindrome(s)); // Output: "bab" or "aba"
    }
}

File: LeetCode_Python/twopoint/LongestPalindromicSubstring.py
class LongestPalindromicSubstring:
    def longestPalindrome(self, s: str) -> str:
        res = ''
        for i in range(len(s)):
            s1 = self.palindrome(s, i, i)
            s2 = self.palindrome(s, i, i + 1)

            res = res if len(res) > len(s1) else s1
            res = res if len(res) > len(s2) else s2
            # candidate = max(s1, s2, key=len)
            # res = max(res, candidate, key=len)

        return res

    def palindrome(self, s: str, left: int, right: int) -> str:
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1

        return s[left + 1: right]


if __name__ == '__main__':
    solution = LongestPalindromicSubstring()
    s1 = "babad"
    print("Expected Output: 'bab' or 'aba'")
    print("Actual Output:", solution.longestPalindrome(s1))

    s2 = "cbbd"
    print("Expected Output: 'bb'")
    print("Actual Output:", solution.longestPalindrome(s2))

    s3 = "a"
    print("Expected Output: 'a'")
    print("Actual Output:", solution.longestPalindrome(s3))
